src/Services/Azure/Authentication/CustomChainedCredential.cs

```diff
@@ -119,4 +135,76 @@ private static DefaultAzureCredential CreateDefaultCredential(string? tenantId)
 
         return new DefaultAzureCredential(defaultCredentialOptions);
     }
+
+    private static TokenCredential? CreateVsCodeBrokerCredential(string? tenantId)
+    {
+        const string vsCodeClientId = "aebc6443-996d-45c2-90f0-388ff96faa56";
+        string userProfile = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
+        string authRecordPath = Path.Combine(userProfile, ".azure", "ms-azuretools.vscode-azureresourcegroups", "authRecord.json");
+        if (!File.Exists(authRecordPath))
+        {
+            // Try .Azure if .azure is not present
+            authRecordPath = Path.Combine(userProfile, ".Azure", "ms-azuretools.vscode-azureresourcegroups", "authRecord.json");
+            if (!File.Exists(authRecordPath))
+            {
+                return null;
+            }
+        }
+
+        AuthenticationRecord? authRecord;
+        try
+        {
+            using var stream = File.OpenRead(authRecordPath);
+            authRecord = AuthenticationRecord.Deserialize(stream);
+        }
+        catch
+        {
+            // Deserialization failed
+            return null;
+        }
+
+        if (authRecord is null)
+            return null;
+
+        // Validate client ID
+        if (!string.Equals(authRecord.ClientId, vsCodeClientId, StringComparison.OrdinalIgnoreCase))
+            return null;
+
+        // Validate tenant ID if present
+        if (!string.IsNullOrEmpty(authRecord.TenantId) && !IsValidTenantId(authRecord.TenantId))
+            return null;
+
+        // Prefer explicit tenantId, else use from auth record
+        string? effectiveTenantId = !string.IsNullOrEmpty(tenantId)
+            ? tenantId
```

Comment:
>tenantId [](http://example.com/codeflow?start=14&length=8)

Should we also be validating `tenantId` with the same level of rigor that we do `authRecord.TenantId` by calling `IsValidTenantId`?

Reply:
not required, as its validated by the respective credentials that consume it.

Reply:
The tenant ID is validated by the Identity SDK. 
we don't have to do it at this layer.